qiskit/qpy/binary_io/value.py

```diff
@@ -290,6 +290,29 @@ def _read_parameter_expression_v3(file_obj, vectors, use_symengine):
 
     payload = file_obj.read(data.expr_size)
     if use_symengine:
+        # This is a horrible hack to workaround the symengine version checking
+        # it's deserialization does. There were no changes to the serialization
+        # format between 0.11 and 0.13 but the deserializer checks that it can't
+        # load across a major or minor version boundary. This works around it
+        # by just lying about the generating version.
+        symengine_version = symengine.__version__.split(".")
+        major = payload[2]
+        minor = payload[3]
+        if int(symengine_version[1]) != minor:
+            if minor not in (11, 13):
+                raise exceptions.QpyError(
+                    f"Incompatible symengine version {major}.{minor} used to generate the QPY "
+                    "payload"
+                )
+            minor_version = int(symengine_version[1])
+            if minor_version not in (11, 13):
+                raise exceptions.QpyError(
+                    f"Incompatible installed symengine version {symengine.__version__} to load "
+                    "this QPY payload"
+                )
```

Comment:
The ordering of this means that we can cope with generation and load from (say) a hypothetical symengine 0.14, but is it worth us risking that?  We could put `symengine>=0.11,<=0.13` in our requirements, potentially.

Not sure how I feel about that, just asking the question.

Reply:
Can we, did I do the logic incorrectly? My intent was that if we have a version mismatch between the payload and the installed library and if either the payload or the installed version is not using 0.11 or 0.13 we'll raise the error. 

There is an edge case here if the major version != 0, but I figured this was unlikely to be an issue (especially with a minor version of `11` or `13`) in any reasonable timescale.

Reply:
Yeah, I'm saying we can attempt to load if there's no payload mismatch and both were 0.14, but given we know there's problems, would we prefer just to forbid 0.14 from entering at all, rather than have it work iff there's a match?

Yeah, I saw the major version thing too, but was just assuming that we'll fix this before that happened.

Reply:
(and by "yeah" I mean I'm agreeing with you - I think you did the logic right)

Reply:
Ah ok, now I get what you're saying. My initial thought was if there is no mismatch I think we're fine as symengine should be totally fine if everything is the same version. I guess it's more a question of whether we think it's sufficiently a risk of foot gunning by generating a payload with 0.14.0 that will fail if someone is using any other version later

Reply:
Yeah, I was kind of hoping you'd have a strong opinion about that last point so I'd be absolved of needing to develop one.  In lieu of anything better, let's stick with the PR as written - upper bounds aren't ideal, and I don't particularly like introducing one within a patch release.

Reply:
Yeah I think that's what we're talking about. That's the only option really is either add a cap to the requirements file in 1.2.3, or we do it in the python code so that `qpy.dump()` fails if you have symenging > 0.14 installed. It might be prudent. I was hesitant to do that in this PR because I was worried about python 3.13 support. But since 0.13 has 3.13 support maybe it's worth it here.

Reply:
I opted to add the cap in: https://github.com/Qiskit/qiskit/pull/13251/commits/7aae9849ec647816a3a790a24ae6e79931122f24

Reply:
For future posterity: Matt and I talked offline before this, and thought that pre-emptively capping is actually the lesser of the two evils.  We don't like introducing an upper bound in a patch release, but we've had upper bounds on `symengine` before, and the hack we're doing to enable compatibility is pretty dangerous, since symengine's deserialisation code can hard crash on malformed payloads.  It feels safer to try and restrict symengine 0.14 from entering (as best as we can), rather than _hoping_ that nothing will break when it's released.